stl/inc/ranges

```diff
@@ -61,108 +61,80 @@ namespace ranges {
-                __r(__l(static_cast<_Ty&&>(_Val)));
+            _NODISCARD constexpr auto operator()(_Ty&& _Val) & noexcept(
+                noexcept(_Right(_Left(_STD forward<_Ty>(_Val))))) requires requires {
+                _Right(_Left(_STD forward<_Ty>(_Val)));
             }
-            { return __r(__l(_STD forward<_Ty>(_Val))); }
+            { return _Right(_Left(_STD forward<_Ty>(_Val))); }
 
             template <class _Ty>
-            _NODISCARD constexpr auto operator()(_Ty&& _Val) const
-                noexcept(noexcept(__r(__l(_STD forward<_Ty>(_Val))))) requires requires {
-                __r(__l(static_cast<_Ty&&>(_Val)));
+            _NODISCARD constexpr auto operator()(_Ty&& _Val) const& noexcept(
+                noexcept(_Right(_Left(_STD forward<_Ty>(_Val))))) requires requires {
+                _Right(_Left(_STD forward<_Ty>(_Val)));
             }
-            { return __r(__l(_STD forward<_Ty>(_Val))); }
+            { return _Right(_Left(_STD forward<_Ty>(_Val))); }
+
+            template <class _Ty>
+            _NODISCARD constexpr auto operator()(_Ty&& _Val) && noexcept(
+                noexcept(_STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val))))) requires requires {
+                _STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val)));
+            }
+            { return _STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val))); }
+
+            template <class _Ty>
+            _NODISCARD constexpr auto operator()(_Ty&& _Val) const&& noexcept(
+                noexcept(_STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val))))) requires requires {
+                _STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val)));
+            }
+            { return _STD move(_Right)(_STD move(_Left)(_STD forward<_Ty>(_Val))); }
         };
 
         template <class _Ty1, class _Ty2>
         _Pipeline(_Ty1, _Ty2) -> _Pipeline<_Ty1, _Ty2>;
+
+        template <class _Left, class _Right>
+            requires _Range_adaptor_closure_object<_Left> //
+                && _Range_adaptor_closure_object<_Right> //
+                && constructible_from<remove_cvref_t<_Left>, _Left> //
+                && constructible_from<remove_cvref_t<_Right>, _Right>
+        constexpr auto operator|(_Left&& __l, _Right&& __r) noexcept(
+            noexcept(_Pipeline{static_cast<_Left&&>(__l), static_cast<_Right&&>(__r)})) {
+            return _Pipeline{static_cast<_Left&&>(__l), static_cast<_Right&&>(__r)};
+        }
+
+        template <class _Left, class _Right>
+            requires _Range_adaptor_closure_object<_Right> && range<_Left>
+        constexpr auto operator|(_Left&& __l, _Right&& __r) noexcept(
+            noexcept(_STD forward<_Right>(__r)(_STD forward<_Left>(__l)))) //
+            requires requires {
+            static_cast<_Right&&>(__r)(static_cast<_Left&&>(__l));
+        }
+        { return _STD forward<_Right>(__r)(_STD forward<_Left>(__l)); }

```

Comment:
I believe (and @CaseyCarter confirmed) that this `operator|` overload should be `_NODISCARD`, because all of `_Pipeline`'s function call operators are.

Casey explained that there are currently no Standard range adaptor closures with side effects whose function call operators would be discardable, none are planned, and such things would be misuses of the design. Thus we can add `_NODISCARD` now, and remove it later if we ever find an example where discarding seems reasonable.